### DISCUSSION TITLE
Is there a way to split `jsonl` into train, dev on the fly

### DISCUSSION THREAD

**Discussion by erogol** [lhotse-speech/lhotse#1018] [Q&A] [answered]
Can I split the JSON file into the train, and dev splits on the fly? Or what would be the suggested way to do that if I didn't do it while creating the JSON?

**Comment by desh2608** [comment]
```python
from lhotse import load_manifest_lazy

cs = load_manifest_lazy("<path-to-file.jsonl.gz>")
N = len(cs)

# if you have a pre-defined train/dev split
train_ids = # list of train cuts
dev_ids = # list of dev cuts
cs_train = cs.subset(cut_ids=train_ids)
cs_dev = cs.subset(cut_ids=dev_ids)
cs_train.to_file("<out-train.jsonl.gz>")
cs_dev.to_file("<out-dev.jsonl.gz>")

# if you don't have pre-defined splits
N_dev = 100
N_train = N - N_dev
cs = cs.shuffle()
cs_train = cs.subset(first=N_train)
cs_dev = cs.subset(last=N_dev)
# save to file as before
```

This should probably do it. You could also just do it using bash tools (such as awk) on the jsonl file.

**Reply by erogol** [reply]
Thanks for the example. 

The issue is I don't know the sample ids. I just want to split a portion of the data as validation at the beginning of the training. I know it is not the best ML practice but you know :)

**Reply by erogol** [reply]
Sorry you already answer that in the code above. Let me try it.

**Reply by erogol** [reply]
I got the following error when I do the 2nd option with ``` cuts.subset(last=1000)```

```TypeError: 'LazyManifestIterator' object is not subscriptable```

I think since lazy returns generator it doesn't know to get to the end of the list.

**Reply by desh2608** [reply]
That's definitely a bug. Should be fixed in https://github.com/lhotse-speech/lhotse/pull/1023.

**Reply by erogol** [reply]
I think even after the fix `N = len(cs)` gets you 

```
NotImplementedError: LazyFilter does not support __len__ because it would require iterating over the whole iterator, which is not possible in a lazy fashion. If you really need to know the length, convert to eager mode first using `.to_eager()`. Note that this will require loading the whole iterator into memory.
```

**Reply by desh2608** [reply]
> I think even after the fix `N = len(cs)` gets you 
> 
> 
> 
> ```
> 
> NotImplementedError: LazyFilter does not support __len__ because it would require iterating over the whole iterator, which is not possible in a lazy fashion. If you really need to know the length, convert to eager mode first using `.to_eager()`. Note that this will require loading the whole iterator into memory.
> 
> ```

Can you show the code you are trying to run?

**Reply by pzelasko** [reply]
You seem to have used .filter somewhere, we don’t support len of filtered manifests, if they were representing sth like 10k hours of speech it would take ages to count since you don’t know how many elements will be filtered out until you iterate.

**Reply by pzelasko** [reply, marked as answer]
As the error says you can call .to_eager() and you’ll be able to use len again
